sql/core/src/main/scala/org/apache/spark/sql/execution/joins/MergeAsOfJoinExec.scala

```diff
@@ -85,31 +92,38 @@ case class MergeAsOfJoinExec(
     var rPrev = rHead.copy()
 
     currLeft._2.map(lHead => {
-        breakable {
-          while (rHead.getInt(0) <= lHead.getInt(0)) {
-            // TODO make index agnostic and check by type (timestamp)
-            var rHeadCopy = rHead.copy()
-            if (currRight._2.hasNext) {
-              rPrev = rHeadCopy.copy()
-              rHeadCopy = currRight._2.next()
-            } else {
-              break
-            }
+      val leftKeyProj = UnsafeProjection.create(Seq(leftOn), left.output)
+      val rightKeyProj = UnsafeProjection.create(Seq(rightOn), right.output)
+      breakable {
+        // Use pointers to determine candidacy of the joining of right rows to left.
+        while (allowExactMatches && rightKeyProj(rHead).getLong(0) <= leftKeyProj(lHead).getLong(0)
+          || !allowExactMatches && rightKeyProj(rHead).getLong(0) < leftKeyProj(lHead).getLong(0)) {
+          var rHeadCopy = rHead.copy()
+          if (currRight._2.hasNext) {
+            rPrev = rHeadCopy.copy()
+            rHeadCopy = currRight._2.next()
```

Comment:
The variable is called `rHeadCopy` however it looks like it's not always a copy. This looks like a potential bug.